scipy/stats/stats.py

```diff
@@ -7340,11 +7347,19 @@ def rankdata(a, method='average'):
     array([ 1,  2,  3,  2])
     >>> rankdata([0, 2, 3, 2], method='ordinal')
     array([ 1,  2,  4,  3])
-    
+    >>> rankdata([[0, 2], [3, 2]]).reshape(2,2)
+    array([[1. , 2.5],
+          [4. , 2.5]])
+    >>> rankdata([[0, 2], [3, 2]], axis = 1)
+    array([[1., 2.],
+          [2., 1.]])
     """
     if method not in ('average', 'min', 'max', 'dense', 'ordinal'):
         raise ValueError('unknown method "{0}"'.format(method))
 
+    if axis is not None:
+        return np.apply_along_axis(rankdata, axis, a, method)
+
```

Comment:
This is the only change to the code, really.

Reply:
Using `apply_along_axis` is fine. Eventually (some future PR), it might be worthwhile trying to pass the `axis` argument into the implementation itself to see if it would be more efficient, but for now, it is hard to argue against the simplicity of a one-liner.

There is one complication that arises when we check the edge cases.  If we're going to handle n-d arrays and the `axis` parameter, it should work with an array that has shape, say, `(3, 0)`.  Currently the function crashes with such an example when `axis=0` (which is a limitation (bug?) of `apply_along_axis`):

```
In [7]: z = np.empty((3, 0))                                                                                                                                          

In [8]: rankdata(z, axis=0)                                                                                                                                           
---------------------------------------------------------------------------
StopIteration                             Traceback (most recent call last)
~/mc37scipy/lib/python3.7/site-packages/numpy/lib/shape_base.py in apply_along_axis(func1d, axis, arr, *args, **kwargs)
    375     try:
--> 376         ind0 = next(inds)
    377     except StopIteration:

StopIteration: 

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-8-55b6424cbf70> in <module>
----> 1 rankdata(z, axis=0)

~/mc37scipy/lib/python3.7/site-packages/scipy-1.5.0.dev0+1a1ab0d-py3.7-macosx-10.7-x86_64.egg/scipy/stats/stats.py in rankdata(a, method, axis)
   7359 
   7360     if axis is not None:
-> 7361         return np.apply_along_axis(rankdata, axis, a, method)
   7362 
   7363     arr = np.ravel(np.asarray(a))

<__array_function__ internals> in apply_along_axis(*args, **kwargs)

~/mc37scipy/lib/python3.7/site-packages/numpy/lib/shape_base.py in apply_along_axis(func1d, axis, arr, *args, **kwargs)
    376         ind0 = next(inds)
    377     except StopIteration:
--> 378         raise ValueError('Cannot apply_along_axis when any iteration dimensions are 0')
    379     res = asanyarray(func1d(inarr_view[ind0], *args, **kwargs))
    380 

ValueError: Cannot apply_along_axis when any iteration dimensions are 0
```
To handle this, an explicit check for `a.size == 0` can be made.    In general, we expect the following from `rankdata`:
* when the input is an n-d array and an `axis` is given, the output has the same shape as the input; and
* the data type of the output is `np.float64` if the method is `'average'`, `np.int_` otherwise;

so I suggest adding a few lines like this:
```
    if axis is not None:
        a = np.asarray(a)
        if a.size == 0:
            # The return values of `normalize_axis_index` are ignored.  The
            # call validates `axis`, even though we won't use it.
            np.core.multiarray.normalize_axis_index(axis, a.ndim)
            dt = np.float64 if method == 'average' else np.int_
            return np.empty(a.shape, dtype=dt)
        return np.apply_along_axis(rankdata, axis, a, method)
```
Technically, `np.core.multiarray.normalize_axis_index` is an internal numpy function that we shouldn't rely on.  If https://github.com/scipy/scipy/pull/11797 is merged, we can use `scipy._lib._util._normalize_axis_index` instead.